src/Algebra/Graph/NonEmpty.hs

```diff
@@ -0,0 +1,563 @@
+-- transpose ('vertex' x)  == 'vertex' x
+-- transpose ('edge' x y)  == 'edge' y x
+-- transpose . transpose == id
+-- transpose . 'path1'     == 'path1'    . 'Data.List.NonEmpty.reverse'
+-- transpose . 'circuit1'  == 'circuit1' . 'Data.List.NonEmpty.reverse'
+-- transpose . 'clique1'   == 'clique1'  . 'Data.List.NonEmpty.reverse'
+-- transpose ('box' x y)   == 'box' (transpose x) (transpose y)
+-- 'edgeList' . transpose  == 'Data.List.sort' . map 'Data.Tuple.swap' . 'edgeList'
+-- @
+transpose :: NonEmptyGraph a -> NonEmptyGraph a
+transpose = foldg1 vertex overlay (flip connect)
+
+-- | Simplify a graph expression. Semantically, this is the identity function,
+-- but it simplifies a given expression according to the laws of the algebra.
+-- The function does not compute the simplest possible expression,
+-- but uses heuristics to obtain useful simplifications in reasonable time.
+-- Complexity: the function performs /O(s)/ graph comparisons. It is guaranteed
+-- that the size of the result does not exceed the size of the given expression.
+--
+-- @
+-- simplify              == id
+-- 'size' (simplify x)     <= 'size' x
+-- simplify 1           '===' 1
+-- simplify (1 + 1)     '===' 1
+-- simplify (1 + 2 + 1) '===' 1 + 2
+-- simplify (1 * 1 * 1) '===' 1 * 1
+-- @
+simplify :: Ord a => NonEmptyGraph a -> NonEmptyGraph a
+simplify = foldg1 Vertex (simple Overlay) (simple Connect)
+
+simple :: Eq g => (g -> g -> g) -> g -> g -> g
+simple op x y
+    | x == z    = x
+    | y == z    = y
+    | otherwise = z
+  where
+    z = op x y
+
+-- | Compute the /Cartesian product/ of graphs.
+-- Complexity: /O(s1 * s2)/ time, memory and size, where /s1/ and /s2/ are the
+-- sizes of the given graphs.
+--
+-- @
+-- box ('path1' $ fromList [0,1]) ('path1' $ fromList "ab") == 'edges1' [ ((0,\'a\'), (0,\'b\'))
+--                                                              , ((0,\'a\'), (1,\'a\'))
+--                                                              , ((0,\'b\'), (1,\'b\'))
+--                                                              , ((1,\'a\'), (1,\'b\')) ]
+-- @
+-- Up to an isomorphism between the resulting vertex types, this operation
+-- is /commutative/, /associative/, /distributes/ over 'overlay', and has
+-- singleton graphs as /identities/. Below @~~@ stands for the equality up to an
+-- isomorphism, e.g. @(x, ()) ~~ x@.
+--
+-- @
+-- box x y               ~~ box y x
+-- box x (box y z)       ~~ box (box x y) z
+-- box x ('overlay' y z)   == 'overlay' (box x y) (box x z)
+-- box x ('vertex' ())     ~~ x
+-- 'transpose'   (box x y) == box ('transpose' x) ('transpose' y)
+-- 'vertexCount' (box x y) == 'vertexCount' x * 'vertexCount' y
+-- 'edgeCount'   (box x y) <= 'vertexCount' x * 'edgeCount' y + 'edgeCount' x * 'vertexCount' y
+-- @
+box :: NonEmptyGraph a -> NonEmptyGraph b -> NonEmptyGraph (a, b)
+box x y = overlays1 xs `overlay` overlays1 ys
+  where
+    xs = fmap (\b -> fmap (,b) x) $ fromJust $ NonEmpty.nonEmpty $ toList y
```

Comment:
should there be `toNonEmpty`?

Reply:
I've added `toNonEmpty` but not sure if we need to expose it. Shall I?

At some point it will become available via `Foldable1`.